# Homework 3
---

# Set up:

- **Do NOT change the name of this jupyter notebook! Otherwise, the autograder will not work as intended!**
- **There will be some cells that appear blank and that you can't type in. That is normal. Just ignore them.**
- **Complete your homework on JupyterHub. Do not copy the file to your local machine.**
- Any files you create for this assignment should be stored in the same directory where you found this notebook. You can also upload your log to that directory when you are finished with the assignment.

   *You don't have to do anything here, just don't CHANGE the path where you save or create files when working on JupyterHub!*
- Once you are finished, you can submit your assignment directly through JupyterHub and we will run the autograder and return feedback to you.
- Once your are finished, please **DO** download a copy of your submitted/{assignment name} folder from JupyterHub to your local machine and commit changes and push to your GitHub repository for your records!
---
Before you turn this assignment in, make sure everything runs as expected. First, **restart the kernel** (in the menubar, select Kernel $\rightarrow$ Restart) and then **run all cells** (in the menubar, select Cell $\rightarrow$ Run All). You can also *Validate* your assignment in the Nbgrader Assignment List to make sure it is correct.

Make sure you fill in any place that says `YOUR CODE HERE` and remove the statement `raise NotImplementedError()` below it.

You are certainly allowed to work together and communicate with your peers to solve this assignment but per UCF Academic Integrity policy, **the work in the cells below MUST be your own unique work to receive credit.**

## Run but otherwise ignore the cell below.

In [44]:
# Runtime setup used by the assignment. Do not edit this cell.
import os
import time
import atexit
from datetime import datetime
import numpy as np

# Automatic Jupyter-session timing log.
# This is different from the student's cumulative .log work-log file.
_session_log_file = "Do_Not_Delete_x43rt2.dat"
_is_grading = os.environ.get("NBGRADER_GRADING") == "1"

if not _is_grading:
    _session_start = time.time()
    _timestamp = datetime.now().isoformat()
    _current_dir = os.getcwd()

    try:
        with open(_session_log_file, "a") as _f:
            _f.write(f"*****IN: {_timestamp} | DIR: {_current_dir}\n")
    except Exception:
        pass

    def _log_session_end():
        _timestamp_out = datetime.now().isoformat()
        try:
            with open(_session_log_file, "a") as _f:
                _f.write(f"*****OUT: {_timestamp_out}\n")
        except Exception:
            pass

    atexit.register(_log_session_end)


## Problem 1
**1.1** We have a body of mass $m$ (kg). In one dimension, when a net force $F_x$ (N) acts on it, according to Newton’s second law, the body will get an acceleration $a_x$ (m/s$^2$) in the same direction. Remember that $F_{NET~x}=ma_x$. Write an expression for each of the following mass and acceleration pairs that calculates $F_{NET~x}$, assigns it to the variable `F` and prints an informative statement about the calculated force. (*You don't need to use arrays of `for` loops just simply calculate F for each pair and then print it.*) 

| $m$ (kg)| $a$ (m/s$^2$)|
| --- | --- |
| 10 | 5 | 
| 15 | 10 | 
| 20 | 5 | 
| 40 | 3 | 

In [45]:
m = 10
a = 15
F = m*a
print(f"When m={m} kg and a={a} m/s^2 then F={F} N.")

m = 15
a = 10
F = m*a
print(f"When m={m} kg and a={a} m/s^2 then F= {F} N.")

m = 20
a = 5
F = m*a
print(f"When m={m} kg and a={a} m/s^2 then F= {F} N.")

m = 40
a = 3
F = m*a
print(f"When m={m} kg and a={a} m/s^2 then F= {F} N.")

When m=10 kg and a=15 m/s^2 then F=150 N.
When m=15 kg and a=10 m/s^2 then F= 150 N.
When m=20 kg and a=5 m/s^2 then F= 100 N.
When m=40 kg and a=3 m/s^2 then F= 120 N.



##### Expected Output:
```
When m=10 kg and a=5 m/s^2 then F=50 N. 
When m=15 kg and a=10 m/s^2 then F=150 N. 
When m=20 kg and a=5 m/s^2 then F=100 N. 
When m=40 kg and a=3 m/s^2 then F=120 N. 
```

In [46]:

# Check your work!
assert F == 120


**1.2** Make a function named `force_to_acceleration_1D` that takes in the mass $m$ (kg) of a body and the magnitude of the net force $F$ (N) applied to it and calculates the acceleration $a ~\mathrm{(m/s^2)}$ (in that order). Remember to write a detailed docstring! Also, remember to write comments throughout your code to explain what you've done. Calculate the acceleration of a body of mass $150$ kg when forces of $100$, $22$ and $450$ are applied to it. Then print an informative statement about each acceleration.

In [47]:
def force_to_acceleration_1D(m,F):
    """Calculates acceleration given mass (m) and net force (F)"""
    a=F/m
    return a
    #Calculates the acceleration using F and m

m = 150 #kg
a = force_to_acceleration_1D(m,100)
print("The acceleration of the",m, "kg mass experiencing a net force of 100N is" ,a, "m/s^2") #calculate acceleration for 100N

m = 150 #kg
a = force_to_acceleration_1D(m,22)
print("The acceleration of the",m, "kg mass experiencing a net force of 100N is" ,a, "m/s^2") #calculate acceleration for 22N

m = 150 #kg
a = force_to_acceleration_1D(m,450)
print("The acceleration of the",m, "kg mass experiencing a net force of 100N is" ,a, "m/s^2") #calculate acceleration for 450N

The acceleration of the 150 kg mass experiencing a net force of 100N is 0.6666666666666666 m/s^2
The acceleration of the 150 kg mass experiencing a net force of 100N is 0.14666666666666667 m/s^2
The acceleration of the 150 kg mass experiencing a net force of 100N is 3.0 m/s^2


##### Expected Call/Output:
```
mass = 150 #kg
a = force_to_acceleration_1D(mass, 100)
print("The acceleration of the ", mass,\
      "kg mass experiencing a net force of 100 N is ",a, "m/s^2." )
```
`The acceleration of the  150 kg mass experiencing a net force of 100 N is  0.6666666666666666 m/s^2. ...`

In [48]:

# Check your work!
assert np.isclose(force_to_acceleration_1D(150, 450), 3.0)


## Problem 2

In 1-D, the average acceleration of a body over the time interval $\Delta t = t_2-t_1$ is $a_{avg} = \frac{\Delta v}{\Delta t}$ where $\Delta v =  v_2-v_1$ is the change in the bodies velocity. 
- Write a function called `acceleration` which takes in `t1,t2,v1,v2` (in that order) and returns a variable which is the average acceleration of the body.
- **Save your function in a separate file called accel.py, not your main homework 3 notebook.** 
- *Give your function a detailed **docstring** and make sure to leave **at least one comment** in the function explaining some step, or units, etc.*

In your main homework 3 notebook, import the function as acceleration (not accel.acceleration or anything else) and calculate the average acceleration of a body for the following times and speeds:

|#|$t_1~\mathrm{(s)}$|$t_2~\mathrm{(s)}$|$v_1~\mathrm{(m/s)}$|$v_2~\mathrm{(m/s)}$|
|--|------- | ------- | ------- | ------- |
|1.|0  | 2  | 10 | 13 |
|2.|10 | 14 | 15 | 30 |
|3.|14 | 16 | 32 | 36 |
|4.|28 | 32 | 42 | 40 |
|5.|0  | 10 | 20 | 22 |

assign them to variables `a1` - `a5` and print them.

In [49]:
from accel import acceleration 

In [50]:
a1 = acceleration(0,2,10,13)
a2 = acceleration(10,14,15,30)
a3 = acceleration(14,16,32,36)
a4 = acceleration(28,32,42,40)
a5 = acceleration(0,10,20,22)

print("accelerations [m/s^2]]:", a1, a2, a3, a4, a5)

accelerations [m/s^2]]: 1.5 3.75 2.0 -0.5 0.2


##### Expected Call/Output:
 
```
a1 = acceleration(0,2,10,13)
a2 = acceleration(10,14,15,30)
a3 = acceleration(14,16,32,36)
a4 = acceleration(28,32,42,40)
a5 = acceleration(0,10,20,22)

print("accelerations [m/s^2]]:", a1, a2, a3, a4, a5)
```

`accelerations [m/s^2]]: 1.5 3.75 2.0 -0.5 0.2`


I checked my values for my a_avg using my phones calculator.

Test case 1: I could have two of the same velocities for two different times, knowing that the resultant acceleration should be 0

Test case 2: I could do times and velocities than i kow would result in the acceleration being 1. For example t0 = 0, t1 = 10, v1 = 0, v2 = 10

- Calculate these numbers some other way to check that they are correct (in general, always check your code in this way). State in comments or markdown below how you checked (calculator, web site, etc.). Also in comments or markdown, name 2 test cases you could use to verify the code works as it should (i.e., what input values could you put into your function for which you know what the output acceleration should be?). Continue to use comments in your code, when appropriate.

## Problem 3
**3.1**
- Create a list called `velocity` with numbers starting from $0$ to $100$ with a step of $10$. 
- Print it. 
- Create a list called `time` with numbers from $0$ to $1000$ with a step of $100$. 
- Print it. 
- In one command, assign the 3rd up to and including the 5th element as well as the 8th element of `velocity` to the list `vels`.
- Print `vels`.

In [51]:
import numpy as np

velocity = np.arange(0,110,10)
print(velocity)

time = np.arange(0,1100,100)
print(time)

vels = velocity[np.r_[2:5,7]]
print(vels)

[  0  10  20  30  40  50  60  70  80  90 100]
[   0  100  200  300  400  500  600  700  800  900 1000]
[20 30 40 70]


**3.2**
Now, using the `acceleration` function you created above, call the function and calculate the acceleration of a body for each of the velocity pairs in `vels` (so the 3rd-4th, 4th-5th, 5th-8th element of `velocity`), use the corresponding elements from `time` and store them in the list `ts`. Store the acceleration values in the list `accels`.

- Print an informative statment containing `accels`.

*Remember that Python starts indexing from 0!*

In [52]:
ts = list(time[[2,3,4,7]])
vels = list(vels)

accels = [
    float(acceleration(ts[0], ts[1], vels[0], vels[1])),
    float(acceleration(ts[1], ts[2], vels[1], vels[2])),
    float(acceleration(ts[2], ts[3], vels[2], vels[3])),] #calculates acceleration for each pair

print("The accelerations in [m/s^2] are:", accels)

The accelerations in [m/s^2] are: [0.1, 0.1, 0.1]


##### Expected Output:

`The accelerations in [m/s^2] are: [0.1, 0.1, 0.1]`


**3.3**
- Change the 5th element of the list `velocity` to $90$ and print the entire list. Does it work? Why/ why not? What if `velocity` was a tuple?

In [53]:
velocity[4] = 90 #changes the velocity in cell 4 to 90 instead of 40
print(velocity)

[  0  10  20  30  90  50  60  70  80  90 100]


Yes this works, because lists in python allow their values to be modified, in other words, they are mutable. If velocity was in a tuple instead of a list, it would result in an error, because tuples are immutable, meaning their values cant be changed after creation.

## Problem 4
- Rewrite your function `force_to_acceleration_1D` from Problem 1 to allow the user to put in the vector net force $\vec{F}_{NET}$ as a list so the function returns the vector acceleration $\vec{a}$ as a list.

- Name your modified function `force_to_acceleration_nD`.

- Would you have to do anything to your original function if `F` was a *numpy array* and you expected the output to be a *numpy array*? Check it!

- Call your function `force_to_acceleration_nD` for a $16 ~\mathrm{kg}$ mass experiencing a net force of $\vec{F}_{NET} = 1.0~ \hat{i} - 2.0~\hat{j} + 3.0~\hat{k} ~\mathrm{N}$ and print the result.

In [54]:
def force_to_acceleration_nD(m, F):
    return [f/m for f in F]

m = 16 #kg
F_net = [1.0, -2.0, 3.0] #N

acceleration_vector = force_to_acceleration_nD(m, F_net)
print(force_to_acceleration_nD(16,[1,-2,3]))

[0.0625, -0.125, 0.1875]


##### Expected Call/Output:
```
print(force_to_acceleration_nD(16,[1,-2,3]))
```
`[0.0625, -0.125, 0.1875]`

In [55]:

# Check your work!
assert np.allclose(
    force_to_acceleration_nD(16, [1, -2, 3]),
    [0.0625, -0.125, 0.1875]
)


---
### Homework Submission
Prepare and submit your homework on JupyterHub.

- Make sure that all the cells run without errors and that your output matches the solution.

  *All your functions will be tested with different inputs to assure that they are working as intended!*
  
- Navigate to the `Home` tab in JupyterNotebook and make sure all additional files (.py files, figures, etc.) are in your assignment folder. 
- Copy your up-to-date `.log` file into your assignment folder.
- Make sure your submission folder is properly named according to the naming conventions.
- After submission, Feel free to download a copy of your submitted assignment folder.  Use GitHub to update and commit your changes on your machine and push to your online GitHub repository.
